Monash-FIT3170/Finding-Neno
# Delete Account and Added Salt to Password

README: 
Re-run setup_db.py to apply the changes to the db.
Your existing accounts will no longer work due to the new salting.

### Delete Account (Merged from #94)
First part of this PR is for the user story: "As a user, I want to be able to delete my account if I no longer want to use the website's services, so that my personal information is no longer stored on the website."
https://app.clickup.com/t/85yxypk9u

Summary:
- Added queries to delete all sightings, reports, and pets associated to the user.
- Also unlinks sightings made by other users on deleting user's missing reports
- Deletes the user from the database.

NOTE: Changes made may be redundant to the user story for deleting pet.
The delete account button location is also subject to change.

To test this: 
1. Create a new user
2. Add a pet, missing report and a new sighting.
3. Use a different account to create a sighting on the new missing report.
4. Go to the profile page and delete the account by completing the confirmation.

### Added Salt to Password
Second part is this adds improvements to password security. Currently, the salt added to the password is a hashed version of the password, meaning that identical passwords will have identical hashed passwords. This PR adds a randomly generated salt to the password before hashing and also adds stronger password requirements when signing up.

Summary:
- Password requirement is stronger (requiring longer password, an uppercase letter, a number, and a symbol)
- A random salt is generated for each user when they sign up or when password is changed
- The salt is stored in the database on a new column 'salt' on users table
- The salt is appended to the password before hashed and stored in db

To test this:
1. Create 2 accounts with the same password. 
2. On TablePlus, these 2 accounts should have different hashed_passwords.
Then test the password requirements, and type different passwords (e.g. 1, abc, Abc1, 12345678Ab) on the password field in the Signup page. See if the validation works properly.
